integrations/helm/releasesync/releasesync.go

```diff
@@ -0,0 +1,349 @@
+/*
+When a Chart release is manually deleted/upgraded/created, the cluster gets out of sync
+with the prescribed state defined by FluxHelmRelease custom resources. The releasesync package
+attemps to bring the cluster back to the prescribed state.
+*/
+package releasesync
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	protobuf "github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/weaveworks/flux/integrations/helm/chartsync"
+	"github.com/weaveworks/flux/integrations/helm/release"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	hapi_release "k8s.io/helm/pkg/proto/hapi/release"
+
+	"github.com/go-kit/kit/log"
+
+	ifv1 "github.com/weaveworks/flux/apis/helm.integrations.flux.weave.works/v1alpha"
+	ifclientset "github.com/weaveworks/flux/integrations/client/clientset/versioned" // kubernetes 1.9
+	helmgit "github.com/weaveworks/flux/integrations/helm/git"
+	chartrelease "github.com/weaveworks/flux/integrations/helm/release"
+)
+
+type action string
+
+const (
+	CustomResourceKind        = "FluxHelmRelease"
+	releaseLagTime            = 30
+	deleteAction       action = "DELETE"
+	installAction      action = "CREATE"
+	upgradeAction      action = "UPDATE"
+)
+
+type ReleaseFhr struct {
+	RelName string
+	FhrName string
+	Fhr     ifv1.FluxHelmRelease
+}
+
+type ReleaseChangeSync struct {
+	logger log.Logger
+	chartsync.Polling
+	kubeClient kubernetes.Clientset
+	ifClient   ifclientset.Clientset
+	release    *chartrelease.Release
+}
+
+func New(
+	logger log.Logger, syncInterval time.Duration, syncTimeout time.Duration,
+	kubeClient kubernetes.Clientset, ifClient ifclientset.Clientset,
+	release *chartrelease.Release) *ReleaseChangeSync {
+
+	return &ReleaseChangeSync{
+		logger:     logger,
+		Polling:    chartsync.Polling{Interval: syncInterval, Timeout: syncTimeout},
+		kubeClient: kubeClient,
+		ifClient:   ifClient,
+		release:    release,
+	}
+}
+
+type customResourceInfo struct {
+	name, releaseName string
+	resource          ifv1.FluxHelmRelease
+	lastUpdated       protobuf.Timestamp
+}
+
+type chartRelease struct {
+	releaseName  string
+	action       action
+	desiredState ifv1.FluxHelmRelease
+}
+
+// Run ... creates a syncing loop monitoring repo chart changes
+func (rs *ReleaseChangeSync) Run(stopCh <-chan struct{}, errc chan error, wg *sync.WaitGroup) {
+	rs.logger.Log("info", "Starting repo charts sync loop")
+
+	wg.Add(1)
+	go func() {
+		defer runtime.HandleCrash()
+		defer wg.Done()
+		defer rs.release.Repo.ReleasesSync.Cleanup()
+
+		time.Sleep(30 * time.Second)
```

Comment:
The sleep is there so that the two syncs dont act at the same time. I shall make the 30ces configurable. Was planning to and forgot.

Reply:
What happens when the syncs happen at the same time? It sounds like that it should not be solved by arbitarily sleeping to offset it.

Reply:
> Was planning to and forgot.

Humans are forgetful. Add `// TODO` comments for yourself so that you notice when reviewing the diff before creating a PR.

If you add the comments after creating a [WIP] PR, add a note to yourself in the description along the lines of _"Note to self: check diff for TODOs before removing [WIP] tag"_

Reply:
Thank you. You are a well of wisdom.

Reply:
@aaron7 
>What happens when the syncs happen at the same time

The chartsync ensures the cluster reflects the state described in the git repo. In particular it syncs (git) changes made to Charts themselves (rather than changes made to overriding values in Custom Resource manifests).

The releasesync looks at all Chart releases in the cluster and, based on deployed FluxHelmRelease resources in the cluster, determines further action. So it is preferable to run the chartsync first to refresh the cluster.

If both syncs run at the same time, the releasesync may be relying on info that will be shortly overriden. Not a big deal as all will catch up next round, but best if the syncing change happens as early as it can. Running both sync in parallel may lead to redundant releases.

> It sounds like that it should not be solved by arbitrarily sleeping to offset it.

In practical terms it is possible to get a good idea about how long the syncing/release takes. If a company runs 30 Charts (we know of one such), it may take about 60 secs which then can be taken into consideration for the sleep delay. But yes, as anything depending on something that is not precisely known, it is not ideal. Please see the Note below.

### Note

I am working on reducing the number of repos used during the helm integration. The releasesync will be performed during the chartsync run, after the charts related sync finishes, so it will almost certainly rely on cluster releases that are reflecting the prescribed state. Should have done that in the first place :thinking:

Reply:
Thanks for the explaination of the syncs. 

> I am working on reducing the number of repos used during the helm integration. The releasesync will be performed during the chartsync run, after the charts related sync finishes

Yep, it sounds like this is the better approach - running the syncs in the same cycle, waiting until one is finished before starting the other.

Reply:
:+1: